src/Raven.Server/Documents/Indexes/Index.cs

```diff
@@ -687,6 +693,49 @@ protected void ExecuteIndexing()
                                     TimeSpentIndexing.Stop();
                                 }
 
+                                if (SideBySideIndex)
+                                {
+                                    var canReplace = true;
+                                    TransactionOperationContext indexContext;
+                                    DocumentsOperationContext documentsOperationContext;
+                                    using (DocumentDatabase.DocumentsStorage.ContextPool.AllocateOperationContext(out documentsOperationContext))
+                                    using (_contextPool.AllocateOperationContext(out indexContext))
+                                    {
+                                        using (indexContext.OpenReadTransaction())
+                                        using (documentsOperationContext.OpenReadTransaction())
+                                        {
+                                            foreach (var collection in Collections)
+                                            {
+                                                var lastProcessedDocEtag =
+                                                    _indexStorage.ReadLastIndexedEtag(indexContext.Transaction, collection);
+                                                if ((MinimumEtagBeforeReplace == null) &&
+                                                    (IsStale(documentsOperationContext, indexContext)) ||
+                                                    ((MinimumEtagBeforeReplace != null) &&
+                                                     (MinimumEtagBeforeReplace < lastProcessedDocEtag)))
+                                                {
+                                                    canReplace = false;
+                                                    break;
+                                                }
+                                            }
+                                        }
+                                        if (canReplace)
+                                        {
+                                            // this can fail if the indexes lock is currently held, so we'll retry
+                                            // however, we might be requested to shutdown, so we want to skip replacing
+                                            // in this case, worst case scenario we'll handle this in the next batch
+                                            while (_cancellationTokenSource.IsCancellationRequested == false)
+                                            {
+                                                if (DocumentDatabase.IndexStore.TryReplaceIndexes(RealName,
```

Comment:
Wouldn't it be better to have IndexReplacer inside IndexStorage that would listen to the BatchCompleted? Will there be an option to replace after certain time?